File: src/test/java/com/github/cybellereaper/wizpets/core/TalentRegistryImplTest.java
package com.github.cybellereaper.wizpets.core;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.github.cybellereaper.wizpets.api.ActivePet;
import com.github.cybellereaper.wizpets.api.StatType;
import com.github.cybellereaper.wizpets.api.talent.PetTalent;
import com.github.cybellereaper.wizpets.api.talent.TalentFactory;
import com.github.cybellereaper.wizpets.core.talent.TalentRegistryImpl;
import java.util.List;
import java.util.Random;
import org.junit.jupiter.api.Test;

class TalentRegistryImplTest {
  private static class DummyTalent implements PetTalent {
    @Override
    public String getId() {
      return "dummy";
    }

    @Override
    public String getDisplayName() {
      return "Dummy";
    }

    @Override
    public String getDescription() {
      return "Does nothing";
    }

    @Override
    public double modifyStat(ActivePet pet, StatType stat, double current) {
      return current + 1;
    }
  }

  private static final class AlternateDummyTalent extends DummyTalent {
    @Override
    public String getDisplayName() {
      return "Alternate Dummy";
    }
  }

  @Test
  void registerAndInstantiateTalents() {
    TalentRegistryImpl registry = new TalentRegistryImpl();
    registry.register((TalentFactory) DummyTalent::new);
    List<PetTalent> talents = registry.instantiate(List.of("dummy"));
    assertEquals(1, talents.size());
    assertEquals("Dummy", talents.get(0).getDisplayName());
    assertThrows(UnsupportedOperationException.class, () -> talents.add(new DummyTalent()));
  }

  @Test
  void registeringDuplicateWithoutReplaceFails() {
    TalentRegistryImpl registry = new TalentRegistryImpl();
    registry.register((TalentFactory) DummyTalent::new);
    assertThrows(
        IllegalArgumentException.class, () -> registry.register((TalentFactory) DummyTalent::new));
  }

  @Test
  void registeringWithReplaceOverridesDescriptor() {
    TalentRegistryImpl registry = new TalentRegistryImpl();
    registry.register((TalentFactory) DummyTalent::new);
    registry.register((TalentFactory) AlternateDummyTalent::new, true);
    assertEquals("Alternate Dummy", registry.get("dummy").getDisplayName());
  }

  @Test
  void rollReturnsIdsFromRegistry() {
    TalentRegistryImpl registry = new TalentRegistryImpl();
    registry.register((TalentFactory) DummyTalent::new);
    List<String> roll = registry.roll(new Random(0), 3);
    assertEquals(3, roll.size());
    assertTrue(roll.stream().allMatch(id -> id.equals("dummy")));
    assertThrows(UnsupportedOperationException.class, () -> roll.add("other"));
  }

  @Test
  void inheritFallsBackToRegistryWhenParentsEmpty() {
    TalentRegistryImpl registry = new TalentRegistryImpl();
    registry.register((TalentFactory) DummyTalent::new);
    List<String> inherited = registry.inherit(List.of(), List.of(), new Random(1), 2);
    assertEquals(2, inherited.size());
    assertTrue(inherited.stream().allMatch(id -> id.equals("dummy")));
  }

  @Test
  void inheritPrefersUniqueTalentsFromParents() {
    TalentRegistryImpl registry = new TalentRegistryImpl();
    registry.register((TalentFactory) DummyTalent::new);
    registry.register(
        () ->
            new PetTalent() {
              @Override
              public String getId() {
                return "other";
              }

              @Override
              public String getDisplayName() {
                return "Other";
              }

              @Override
              public String getDescription() {
                return "Other talent";
              }
            });
    List<String> inherited = registry.inherit(List.of("dummy"), List.of("other"), new Random(5), 2);
    assertTrue(inherited.contains("dummy"));
    assertTrue(inherited.contains("other"));
  }
}

File: src/test/kotlin/com/github/cybellereaper/wizpets/core/TalentRegistryImplTest.kt
package com.github.cybellereaper.wizpets.core

import com.github.cybellereaper.wizpets.api.ActivePet
import com.github.cybellereaper.wizpets.api.StatType
import com.github.cybellereaper.wizpets.api.talent.PetTalent
import com.github.cybellereaper.wizpets.api.talent.TalentFactory
import com.github.cybellereaper.wizpets.core.talent.TalentRegistryImpl
import kotlin.random.Random
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertTrue

private class DummyTalent : PetTalent {
    override val id: String = "dummy"
    override val displayName: String = "Dummy"
    override val description: String = "Does nothing"
    override fun modifyStat(pet: ActivePet, stat: StatType, current: Double): Double = current + 1
}

class TalentRegistryImplTest {
    @Test
    fun `register and instantiate talents`() {
        val registry = TalentRegistryImpl()
        registry.register(TalentFactory { DummyTalent() })
        val talents = registry.instantiate(listOf("dummy"))
        assertEquals(1, talents.size)
        assertEquals("Dummy", talents.first().displayName)
    }

    @Test
    fun `registering duplicate without replace fails`() {
        val registry = TalentRegistryImpl()
        registry.register(TalentFactory { DummyTalent() })
        assertFailsWith<IllegalArgumentException> {
            registry.register(TalentFactory { DummyTalent() })
        }
    }

    @Test
    fun `roll returns ids from registry`() {
        val registry = TalentRegistryImpl()
        registry.register(TalentFactory { DummyTalent() })
        val roll = registry.roll(Random(0), 3)
        assertTrue(roll.all { it == "dummy" })
    }

    @Test
    fun `inherit falls back to registry when parents empty`() {
        val registry = TalentRegistryImpl()
        registry.register(TalentFactory { DummyTalent() })
        val inherited = registry.inherit(emptyList(), emptyList(), Random(1), count = 2)
        assertEquals(2, inherited.size)
        assertTrue(inherited.all { it == "dummy" })
    }
}
